--- origsrc/gftp-2.0.19/configure.in	2008-11-29 08:18:14.000000000 -0600
+++ src/gftp-2.0.19/configure.in	2009-01-01 02:34:58.362500000 -0600
@@ -148,7 +148,7 @@
 USE_READLINE="yes"
 READLINE_LIBS=""
 if test "x$enable_textport" = "xyes" ; then
-  GFTP_TEXT="gftp-text"
+  GFTP_TEXT="gftp-text\$(EXEEXT)"
 
   AC_MSG_CHECKING([whether to use readline])
   AC_ARG_WITH(readline,
@@ -243,17 +243,17 @@
 
 if test "x$enable_gtkport" = "xyes" ; then
   if test $found_glib20 = 1 -a "x$enable_gtk20" = "xyes" ; then
-    PKG_CHECK_MODULES(GTK, gtk+-2.0 >= 2.0.0, GFTP_GTK=gftp-gtk, AC_MSG_ERROR(You have GLIB 2.0 installed but I cannot find GTK+ 2.0. Run configure with --disable-gtk20 or install GTK+ 2.0))
+    PKG_CHECK_MODULES(GTK, gtk+-2.0 >= 2.0.0, GFTP_GTK="gftp-gtk\$(EXEEXT)", AC_MSG_ERROR(You have GLIB 2.0 installed but I cannot find GTK+ 2.0. Run configure with --disable-gtk20 or install GTK+ 2.0))
   fi
 
   if test "x$GFTP_GTK" = "x" ; then
-    AM_PATH_GTK(1.2.3, GFTP_GTK=gftp-gtk, AC_MSG_WARN(gFTP needs GTK+ 1.2.3 or higher for the graphical version. Only building the text version))
+    AM_PATH_GTK(1.2.3, GFTP_GTK="gftp-gtk\$(EXEEXT)", AC_MSG_WARN(gFTP needs GTK+ 1.2.3 or higher for the graphical version. Only building the text version))
     GTHREAD_LIBS="-lgthread"
   else
     GTHREAD_LIBS="-lgthread-2.0"
   fi
 
-  if test "x$GFTP_GTK" = xgftp-gtk; then
+  if test "x$GFTP_GTK" != x; then
     case $host in
       *-hpux*)
         AC_CHECK_LIB(rt, sem_init)
