--- origsrc/gftp-2.0.19/configure.in	2008-11-29 08:18:14.000000000 -0600
+++ src/gftp-2.0.19/configure.in	2009-01-01 02:34:58.362500000 -0600
@@ -79,6 +79,51 @@
 #include <sys/sem.h>
 ])
 
+# Checks for typedefs, structures, and compiler characteristics.
+AC_CHECK_MEMBER(struct dirent.d_namlen,AC_DEFINE(HAVE_DIRENT_NAMLEN,1,[Define if struct dirent has member d_namlen]),,[#include <sys/types.h>
+#include <dirent.h>
+])
+AC_CHECK_MEMBER(struct dirent.d_fileno,AC_DEFINE(HAVE_DIRENT_FILENO,1,[Define if struct dirent has member d_fileno]),,[#include <sys/types.h>
+#include <dirent.h>
+])
+AC_CHECK_MEMBER(struct dirent.d_type,AC_DEFINE(HAVE_DIRENT_TYPE,1,[Define if struct dirent has member d_type]),,[#include <sys/types.h>
+#include <dirent.h>
+])
+AC_CHECK_TYPE(union semun, ,AC_DEFINE(_SEM_SEMUN_UNDEFINED,1,[Define if you do not have semun in sys/sem.h]),
+[#include <sys/types.h>
+#include <sys/ipc.h>
+#include <sys/sem.h>
+])
+
+# Checks for library functions.
+AC_CHECK_FUNCS(lockf semop shmget)
+
+AC_ARG_WITH(lockprefix,AS_HELP_STRING([--with-lockprefix=path],[Set lock prefix path to (default /tmp/.FSPL)]))
+AC_ARG_WITH([locking],AS_HELP_STRING([--with-locking=none/semop/lockf],[Set client locking type (default autodetected)]))
+
+#Parse user locking choice
+if test "x$with_locking" = "xno" -o "x$with_locking" = "xnone" ; then
+    AC_DEFINE(FSP_NOLOCKING, 1, [No FSP locking])
+    AC_MSG_NOTICE(locking disabled by user)
+elif test "x$with_locking" = "xlockf" -a "x$ac_cv_func_lockf" = "xyes"; then
+    AC_DEFINE(FSP_USE_LOCKF, 1, [FSP uses lockf])
+elif test "x$with_locking" = "xsemop" -a "x$ac_cv_func_semop" = "xyes" -a "x$ac_cv_func_shmget" = "xyes";then
+    AC_DEFINE(FSP_USE_SHAREMEM_AND_SEMOP, 1, [FSP uses shmget and semop])
+elif test "x$ac_cv_func_semop" = "xyes" -a "x$ac_cv_func_shmget" = "xyes"; then
+    AC_DEFINE(FSP_USE_SHAREMEM_AND_SEMOP, 1, [FSP uses shmget and semop])
+elif test "x$ac_cv_func_lockf" = "xyes"; then
+    AC_DEFINE(FSP_USE_LOCKF, 1, [FSP uses lockf])
+else
+    AC_DEFINE(FSP_NOLOCKING, 1, [No FSP locking])
+    AC_MSG_NOTICE([no suitable locking method detected])
+fi
+
+#locking prefix
+if test "x$with_lockprefix" != "xno" -a "x$with_lockprefix" != "xyes" -a "x$with_lockprefix" != "x" ; then
+    AC_DEFINE_UNQUOTED(FSP_KEY_PREFIX, "$with_lockprefix" , [FSP locking prefix])
+fi
+
+
 if test "x$enable_gtk20" = "xyes" ; then
   PKG_CHECK_MODULES(GLIB, glib-2.0 >= 2.0.0, found_glib20=1, found_glib20=0)
 else
--- origsrc/gftp-2.0.19/lib/fsp.c	2007-03-15 21:48:08.000000000 -0500
+++ src/gftp-2.0.19/lib/fsp.c	2009-01-01 02:34:58.409375000 -0600
@@ -21,7 +21,6 @@
 #include "gftp.h"
 static const char cvsid[] = "$Id: fsp.c 903 2007-03-16 02:44:07Z masneyb $";
 
-#define FSP_USE_SHAREMEM_AND_SEMOP 1
 #include "fsplib/fsplib.h"
 #include "fsplib/lock.h"
 
--- origsrc/gftp-2.0.19/lib/fsplib/Makefile.am	2008-11-29 08:18:05.000000000 -0600
+++ src/gftp-2.0.19/lib/fsplib/Makefile.am	2009-01-01 02:34:58.425000000 -0600
@@ -2,5 +2,4 @@
 
 noinst_LIBRARIES = libfsp.a
 libfsp_a_SOURCES=fsplib.c lock.c
-INCLUDES=-DFSP_USE_SHAREMEM_AND_SEMOP=1
 noinst_HEADERS=fsplib.h lock.h
--- origsrc/gftp-2.0.19/lib/fsplib/fsplib.c	2008-11-29 06:47:18.000000000 -0600
+++ src/gftp-2.0.19/lib/fsplib/fsplib.c	2009-01-01 02:40:55.909375000 -0600
@@ -12,6 +12,10 @@
                      This is a free software.  Be creative.
                     Let me know of any bugs and suggestions.
 */                  
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/time.h>
@@ -29,6 +33,10 @@
 #include <stdint.h>
 #endif
 
+#ifndef HAVE_GETADDRINFO
+#include <getaddrinfo.h>
+#endif
+
 #include "fsplib.h"
 #include "lock.h"
 
@@ -634,7 +642,9 @@
 #ifdef HAVE_DIRENT_FILENO
     entry->d_fileno = 10;
 #endif    
+#ifdef HAVE_DIRENT_RECLEN
     entry->d_reclen = fentry.reclen;
+#endif
     strncpy(entry->d_name,fentry.name,MAXNAMLEN);
 
     if (fentry.namlen >= MAXNAMLEN)
--- origsrc/gftp-2.0.19/lib/fsplib/lock.c	2008-11-29 06:47:18.000000000 -0600
+++ src/gftp-2.0.19/lib/fsplib/lock.c	2009-01-01 02:38:33.456250000 -0600
@@ -1,3 +1,7 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include <string.h>
 #include <stdio.h>
 #include "lock.h"
